misc/perf_compare.py

```diff
@@ -0,0 +1,145 @@
+"""Compare performance of mypyc-compiled mypy between one or more commits/branches.
+
+Simple usage:
+
+  python misc/perf_compare.py my-branch master ...
+
+What this does:
+
+ * Create a temp clone of the mypy repo for each target commit to measure
+ * Checkout a target commit in each of the clones
+ * Compile mypyc in each of the clones *in parallel*
+ * Create another temp clone of the mypy repo as the code to check
+ * Self check with each of the compiled mypys N times
+ * Report the average runtimes and relative performance
+ * Remove the temp clones
+"""
+
+from __future__ import annotations
+
+import argparse
+import glob
+import os
+import random
+import shutil
+import statistics
+import subprocess
+import sys
+import threading
+import time
+
+
+def heading(s: str) -> None:
+    print()
+    print(f"=== {s} ===")
+    print()
+
+
+def build_mypy(target_dir: str) -> None:
+    env = os.environ.copy()
+    env["CC"] = "clang"
+    env["MYPYC_OPT_LEVEL"] = "2"
+    cmd = [sys.executable, "setup.py", "--use-mypyc", "build_ext", "--inplace"]
+    subprocess.run(cmd, env=env, check=True, cwd=target_dir)
+
+
+def clone(target_dir: str, commit: str | None) -> None:
+    heading(f"Cloning mypy to {target_dir}")
+    repo_dir = os.getcwd()
+    if os.path.isdir(target_dir):
+        print(f"{target_dir} exists: deleting")
+        shutil.rmtree(target_dir)
+    subprocess.run(["git", "clone", repo_dir, target_dir], check=True)
+    if commit:
+        subprocess.run(["git", "checkout", commit], check=True, cwd=target_dir)
+
+
+def run_benchmark(compiled_dir: str, check_dir: str) -> float:
+    cache_dir = os.path.join(compiled_dir, ".mypy_cache")
+    if os.path.isdir(cache_dir):
+        shutil.rmtree(cache_dir)
+    env = os.environ.copy()
+    env["PYTHONPATH"] = os.path.abspath(compiled_dir)
+    abschk = os.path.abspath(check_dir)
+    cmd = [
+        sys.executable,
+        "-m",
+        "mypy",
+        "--config-file",
+        os.path.join(abschk, "mypy_self_check.ini"),
+    ]
+    cmd += glob.glob(os.path.join(abschk, "mypy/*.py"))
+    cmd += glob.glob(os.path.join(abschk, "mypy/*/*.py"))
+    t0 = time.time()
```

Comment:
Is it worth using `time.perf_counter()` or `time.perf_counter_ns()` instead of `time.time()`?

Reply:
Probably doesn't make much difference in practice, since the measured runtimes will be several seconds.